PopPUNK/lineage_clustering.py

```diff
@@ -0,0 +1,680 @@
+                                                                R,
+                                                                qlist,
+                                                                neighbours,
+                                                                lineage_clustering,
+                                                                null_cluster_value)
+    else:
+        neighbours = generate_nearest_neighbours(distances,
+                                                row_labels,
+                                                isolate_list,
+                                                R)
+
+    # run clustering
+    lineage_index = 1
+    while null_cluster_value in lineage_clustering.values():
+
+        # get seed isolate based on minimum pairwise distances
+        seed_isolate = get_seed_isolate(lineage_clustering,
+                                        row_labels,
+                                        distances,
+                                        null_cluster_value,
+                                        lineage_index,
+                                        lineage_seed)
+        lineage_seed[lineage_index] = seed_isolate
+        
+        # seed isolate is None if a previously-existing cluster has been overwritten
+        # in which case pass over the lineage to keep nomenclature consistent
+        if seed_isolate is not None:
+        
+            # record status of seed isolate
+            lineage_clustering[seed_isolate] = lineage_index
+            
+            # get information for lineage clustering
+            lineage_clustering_information = get_lineage_clustering_information(seed_isolate,
+                                                                                row_labels,
+                                                                                distances)
+                                                                                
+            # cluster the lineages
+            lineage_clustering = run_lineage_clustering(lineage_clustering,
+                                                        lineage_clustering_information,
+                                                        neighbours,
+                                                        R,
+                                                        lineage_index,
+                                                        seed_isolate,
+                                                        previous_lineage_clustering,
+                                                        null_cluster_value)
+                                                        
+        # increment index for next lineage
+        lineage_index = lineage_index + 1
+
+    # return clustering
+    return lineage_clustering, lineage_seed, neighbours, previous_lineage_clustering
+
+def calculateQueryDistances(dbFuncs, rlist, qfile, kmers, estimated_length,
```

Comment:
If this functionality cannot already be replicated, sketchlib.py and/or mash.py is a better place for this function

Reply:
Moved to sketchlib and used in main and network.py in 8e0875f.